pkg/reconciler/pipelinerun/pipelinerun.go

```diff
@@ -462,18 +473,41 @@ func (c *Reconciler) reconcile(ctx context.Context, pr *v1beta1.PipelineRun) err
 
 // runNextSchedulableTask gets the next schedulable Tasks from the dag based on the current
 // pipeline run state, and starts them
-func (c *Reconciler) runNextSchedulableTask(ctx context.Context, pr *v1beta1.PipelineRun, d *dag.Graph, pipelineState resources.PipelineRunState, as artifacts.ArtifactStorageInterface) error {
+// after all DAG tasks are done, its responsible for scheduling final tasks and start executing them
+func (c *Reconciler) runNextSchedulableTask(ctx context.Context, pr *v1beta1.PipelineRun, d *dag.Graph, dfinally *dag.Graph, pipelineState resources.PipelineRunState, as artifacts.ArtifactStorageInterface) error {
 
 	logger := logging.FromContext(ctx)
 	recorder := controller.GetEventRecorder(ctx)
 
-	candidateTasks, err := dag.GetSchedulable(d, pipelineState.SuccessfulPipelineTaskNames()...)
-	if err != nil {
-		logger.Errorf("Error getting potential next tasks for valid pipelinerun %s: %v", pr.Name, err)
-		return nil
+	var nextRprts []*resources.ResolvedPipelineRunTask
+	// When the pipeline run is stopping, we don't schedule any new task and only
+	// wait for all running tasks to complete and report their status
+	// after running tasks are done, check if there are any final task which needs to be scheduled
+	if pipelineState.IsStopping(d) {
+		// GetFinalTasks returns a list of final tasks which have not started executing
+		// GetFinalTasks returns tasks only after all DAG tasks have finished executing successfully
+		// or a DAG task might have failed and none parallel tasks are running
+		nextRprts = pipelineState.GetFinalTasks(d, dfinally)
+	} else {
+		// list of candidates is derived based on successfully finished tasks and/or skipped tasks
+		// a task is considered as candidate if its all parents have finished executing successfully
+		// this list of candidates is initialized to DAG root nodes to start pipeline execution
+		candidateTasks, err := dag.GetSchedulable(d, pipelineState.SuccessfulOrSkippedDAGTasks(d)...)
+		if err != nil {
+			logger.Errorf("Error getting potential next tasks for valid pipelinerun %s: %v", pr.Name, err)
+			return nil
+		}
+		// GetNextTasks returns a list of tasks which should be executed next i.e.
+		// a list of tasks from pipelineState which does not have any taskRun or a list of failed tasks which haven't exhausted their retries
+		// Pipeline execution continues until GetNextTasks does not return any new pipeline task to execute
+		nextRprts = pipelineState.GetNextTasks(candidateTasks)
+
+		// execution queue is empty, check if any final tasks are specified and needs to be pushed into execution queue
+		if len(nextRprts) == 0 {
```

Comment:
I think there is a problem here.

`GetPipelineConditionStatus` is used to evaluate what the current condition of a pipeline. It does that by looking at *all* pipeline tasks and making sure that there is nothing else that could be executed at the time.

The code here instead uses the fact that `GetNextTasks` returns an empty list to decide that the pipeline run has completed its work, which it does not seem like valid assumption. 
Let's consider the case where all the tasks returned by `GetSchedulable` are running for instance. The list of next tasks will be empty, because there's nothing new to be scheduled during this reconcile cycle, however there may me more tasks to be executed once the running one are complete.

Reply:
None of the new unit tests introduced has taskruns in "running" state, they are always either failed or succeeded or not started. I think we need to have coverage for cases with running taskruns as well, because that's when the `GetNextTasks` may return an empty list.

Reply:
`GetFinalTasks` is called when `GetNextTasks` returns an empty list but `GetFinalTasks` does not return any task unless DAG tasks are done ([isDAGTasksDone](https://github.com/tektoncd/pipeline/pull/2661/files#diff-4e301064fef70f3c0db810267c0959f2R281)). But there was a problem in `GetFinalTasks` which was returning final tasks if any dag task fails or dag tasks are done. Thanks a bunch for creating a unit test over slack 😲 , added it [here](https://github.com/tektoncd/pipeline/pull/2661/files#diff-f7c8532cfe79319c0fff711a4979a6d6R3567) and updated `GetNextTasks` to only return final tasks after DAG tasks are done.